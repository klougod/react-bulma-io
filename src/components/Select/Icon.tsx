import React from 'react'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'

export const Icon = ({ icon }: any) => (
  <span className='icon is-small is-left'>
    <FontAwesomeIcon icon={icon} />
  </span>
)

Icon.defaultProps = {
  icon: null
}

Icon.propTypes = {
  icon: PropTypes.any
}

export default Icon
