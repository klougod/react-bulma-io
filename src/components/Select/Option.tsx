import React, { ReactNode, OptionHTMLAttributes } from 'react'

interface OptionProps extends OptionHTMLAttributes<HTMLOptionElement> {
  text?: ReactNode
}

export const Option = ({ text, ...rest }: OptionProps) => (
  <option {...rest}>{text}</option>
)

Option.defaultProps = {
  text: '',
  value: '',
  className: '',
  disabled: false
}

export default Option
