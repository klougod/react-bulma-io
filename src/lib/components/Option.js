import React from 'react'
import PropTypes from 'prop-types'

export const Option = ({ value, text, className }) => (
  <option className={className} value={value}>
    {text}
  </option>
)

Option.defaultProps = {
  text: '',
  value: '',
  className: ''
}

Option.propTypes = {
  text: PropTypes.string,
  value: PropTypes.any,
  className: PropTypes.string
}

export default Option