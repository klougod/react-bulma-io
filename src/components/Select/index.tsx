import React, { Children, ReactElement, ReactNode } from 'react'

import Option from './Option'
import Icon from '../Icon'

const optionType = (<Option />).type

interface SelectProps {
  icon?: any
  className?: string
  onChange?: () => void
  placeholder?: string
  children?: ReactNode
}

interface SelectWrapperProps extends SelectProps {
  noWrapper?: boolean
}

const SelectNoWrapper = ({
  className,
  placeholder,
  onChange,
  children
}: SelectProps) => (
  <div
    className={`select ${className}`}
    placeholder={placeholder}
    onChange={onChange}
  >
    <select>
      {Children.toArray(children).filter(
        (c: ReactElement) => c.type === optionType
      )}
    </select>
  </div>
)

const withWrapper = (Element: any) => ({
  icon,
  ...selectProps
}: SelectProps) => {
  return (
    <div className='field'>
      <div className={`control ${icon && 'has-icons-left'}`}>
        {icon && <Icon icon={icon} />}
        <Element {...selectProps} />
      </div>
    </div>
  )
}

export const Select = ({
  noWrapper,
  children,
  ...selectProps
}: SelectWrapperProps) => {
  const SelectComponent = noWrapper
    ? SelectNoWrapper
    : withWrapper(SelectNoWrapper)
  return <SelectComponent {...selectProps}>{children}</SelectComponent>
}

Select.defaultProps = {
  value: '',
  icon: null,
  placeholder: '',
  className: '',
  onChange: null,
  noWrapper: false
}

export default Select
