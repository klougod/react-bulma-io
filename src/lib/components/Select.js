import React, { Children } from 'react'
import PropTypes from 'prop-types'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import Option from './Option';

const optionType = (<Option />).type

export const Select = ({ icon, placeholder, className, onChange, children, noWrapper }) => {

  const Icon = () => (
    <span className="icon is-small is-left">
      <FontAwesomeIcon icon={icon} />
    </span>
  )

  return (
    noWrapper ?
    <div className={`select ${className}`} placeholder={placeholder} onChange={onChange}>
      <select>
        { Children.toArray(children).filter(c => c.type === optionType) }
      </select>
    </div> :
    <div className='field'>
      <div className={`control ${icon && 'has-icons-left'}`}>
        {icon && <Icon />}
        <div className={`select ${className}`} placeholder={placeholder} onChange={onChange}>
          <select>
            { Children.toArray(children).filter(c => c.type === optionType) }
          </select>
        </div>
      </div>
    </div>
  );
}

Select.defaultProps = {
  value: '',
  icon: null,
  placeholder: '',
  className: '',
  onChange: null,
  noWrapper: false
}

Select.propTypes = {
  value: PropTypes.string,
  placeholder: PropTypes.string,
  icon: PropTypes.any,
  className: PropTypes.string,
  onChange: PropTypes.func,
  noWrapper: PropTypes.bool
}

export default Select