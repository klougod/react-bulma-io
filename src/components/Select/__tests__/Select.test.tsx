import React from 'react'
import Enzyme from 'enzyme'
import { faAdjust } from '@fortawesome/free-solid-svg-icons'

import Select from '../'
import Option from '../Option'

describe('Select component test', function () {
  it('should render without crash with wrapper', function () {
    const selectMock = Enzyme.shallow(<Select />)
    expect(selectMock).toMatchSnapshot()
  })

  it('should render without crash with wrapper', function () {
    const selectMockProps = { noWrapper: true }
    const selectMock = Enzyme.shallow(
      <Select {...selectMockProps}>
        <Option />
        <Option />
        <Option />
      </Select>
    )
    expect(selectMock).toMatchSnapshot()
  })

  it('should render without crash with children elements', function () {
    const selectMockProps = { icon: faAdjust }
    const selectMock = Enzyme.shallow(
      <Select {...selectMockProps}>
        <Option />
        <Option />
        <Option />
      </Select>
    )
    expect(selectMock).toMatchSnapshot()
  })
})
