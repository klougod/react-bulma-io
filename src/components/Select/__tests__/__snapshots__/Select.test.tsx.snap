// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component test should render without crash with children elements 1`] = `
<Component
  className=""
  icon={
    Object {
      "icon": Array [
        512,
        512,
        Array [],
        "f042",
        "M8 256c0 136.966 111.033 248 248 248s248-111.034 248-248S392.966 8 256 8 8 119.033 8 256zm248 184V72c101.705 0 184 82.311 184 184 0 101.705-82.311 184-184 184z",
      ],
      "iconName": "adjust",
      "prefix": "fas",
    }
  }
  noWrapper={false}
  onChange={null}
  placeholder=""
  value=""
>
  <Component
    className=""
    icon={
      Object {
        "icon": Array [
          512,
          512,
          Array [],
          "f042",
          "M8 256c0 136.966 111.033 248 248 248s248-111.034 248-248S392.966 8 256 8 8 119.033 8 256zm248 184V72c101.705 0 184 82.311 184 184 0 101.705-82.311 184-184 184z",
        ],
        "iconName": "adjust",
        "prefix": "fas",
      }
    }
    onChange={null}
    placeholder=""
    value=""
  >
    <div
      className="field"
    >
      <div
        className="control has-icons-left"
      >
        <Component
          icon={
            Object {
              "icon": Array [
                512,
                512,
                Array [],
                "f042",
                "M8 256c0 136.966 111.033 248 248 248s248-111.034 248-248S392.966 8 256 8 8 119.033 8 256zm248 184V72c101.705 0 184 82.311 184 184 0 101.705-82.311 184-184 184z",
              ],
              "iconName": "adjust",
              "prefix": "fas",
            }
          }
        >
          <span
            className="icon is-small is-left"
          >
            <FontAwesomeIcon
              border={false}
              className=""
              fixedWidth={false}
              flip={null}
              icon={
                Object {
                  "icon": Array [
                    512,
                    512,
                    Array [],
                    "f042",
                    "M8 256c0 136.966 111.033 248 248 248s248-111.034 248-248S392.966 8 256 8 8 119.033 8 256zm248 184V72c101.705 0 184 82.311 184 184 0 101.705-82.311 184-184 184z",
                  ],
                  "iconName": "adjust",
                  "prefix": "fas",
                }
              }
              inverse={false}
              listItem={false}
              mask={null}
              pull={null}
              pulse={false}
              rotation={null}
              size={null}
              spin={false}
              swapOpacity={false}
              symbol={false}
              title=""
              transform={null}
            >
              <svg
                aria-hidden="true"
                className="svg-inline--fa fa-adjust fa-w-16 "
                data-icon="adjust"
                data-prefix="fas"
                focusable="false"
                role="img"
                style={Object {}}
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M8 256c0 136.966 111.033 248 248 248s248-111.034 248-248S392.966 8 256 8 8 119.033 8 256zm248 184V72c101.705 0 184 82.311 184 184 0 101.705-82.311 184-184 184z"
                  fill="currentColor"
                  style={Object {}}
                />
              </svg>
            </FontAwesomeIcon>
          </span>
        </Component>
        <SelectNoWrapper
          className=""
          onChange={null}
          placeholder=""
          value=""
        >
          <div
            className="select "
            onChange={null}
            placeholder=""
          >
            <select>
              <Component
                className=""
                disabled={false}
                key=".0"
                text=""
                value=""
              >
                <option
                  className=""
                  disabled={false}
                  value=""
                />
              </Component>
              <Component
                className=""
                disabled={false}
                key=".1"
                text=""
                value=""
              >
                <option
                  className=""
                  disabled={false}
                  value=""
                />
              </Component>
              <Component
                className=""
                disabled={false}
                key=".2"
                text=""
                value=""
              >
                <option
                  className=""
                  disabled={false}
                  value=""
                />
              </Component>
            </select>
          </div>
        </SelectNoWrapper>
      </div>
    </div>
  </Component>
</Component>
`;

exports[`Select component test should render without crash with wrapper 1`] = `
<Component
  className=""
  icon={null}
  noWrapper={false}
  onChange={null}
  placeholder=""
  value=""
>
  <Component
    className=""
    icon={null}
    onChange={null}
    placeholder=""
    value=""
  >
    <div
      className="field"
    >
      <div
        className="control null"
      >
        <SelectNoWrapper
          className=""
          onChange={null}
          placeholder=""
          value=""
        >
          <div
            className="select "
            onChange={null}
            placeholder=""
          >
            <select />
          </div>
        </SelectNoWrapper>
      </div>
    </div>
  </Component>
</Component>
`;

exports[`Select component test should render without crash with wrapper 2`] = `
<Component
  className=""
  icon={null}
  noWrapper={true}
  onChange={null}
  placeholder=""
  value=""
>
  <SelectNoWrapper
    className=""
    icon={null}
    onChange={null}
    placeholder=""
    value=""
  >
    <div
      className="select "
      onChange={null}
      placeholder=""
    >
      <select>
        <Component
          className=""
          disabled={false}
          key=".0"
          text=""
          value=""
        >
          <option
            className=""
            disabled={false}
            value=""
          />
        </Component>
        <Component
          className=""
          disabled={false}
          key=".1"
          text=""
          value=""
        >
          <option
            className=""
            disabled={false}
            value=""
          />
        </Component>
        <Component
          className=""
          disabled={false}
          key=".2"
          text=""
          value=""
        >
          <option
            className=""
            disabled={false}
            value=""
          />
        </Component>
      </select>
    </div>
  </SelectNoWrapper>
</Component>
`;
